File: 【单调栈】2023Q1A-删除重复数字后的最大数字.cpp
/*
 * @hw id=2023Q1A lang=cpp
 *
 * 删除重复数字后的最大数字
 */

// @hw code=start
#include<iostream>
#include<vector>
#include<map>
#include<set>

using namespace std;

int main() {
    string s;
    cin >> s;
    map<char, int> cnt;
    for (char c : s) {
        cnt[c] = cnt.find(c)!=cnt.end() ? cnt[c]+1 : 1;
    }
    set<char> used;
    vector<char> stack;
    for(char c : s) {
        if(used.find(c) != used.end()) {
            cnt[c]--;
        } else {
            while(stack.size()>0 && c > stack.back() && cnt[stack.back()]>1) {
                cnt[stack.back()]--;
                used.erase(stack.back());
                stack.pop_back();
            }
            stack.push_back(c);
            used.insert(c);
        }
    }
    for(char c : stack) {
        cout << c;
    }
    cout << endl;
}
// @hw code=end

File: 【单调栈】2023Q1A-删除重复数字后的最大数字.py
#
# @hw id=2023Q1A lang=python3
#
# 删除重复数字后的最大数字
#

# @hw code=start
s = list(input())
cnt = {}

for c in s:
    if c in cnt:
        cnt[c] += 1
    else:
        cnt[c] = 1

used = set()
stack = list()

for c in s:
    if c in used:
        cnt[c] -= 1
    else:
        while len(stack) > 0 and c > stack[-1] and cnt[stack[-1]] > 1:
            oldc = stack.pop()
            cnt[oldc] -= 1
            used.remove(oldc)
        used.add(c)
        stack.append(c)

print("".join(stack))
# @hw code=end